### DISCUSSION TITLE
Command line data, in code data

### DISCUSSION THREAD

**Discussion by K20shores** [fortran-lang/fortls#379] [Q&A]
Is it possible to use this directly in code and not as a language server?

**Comment by gnikit** [comment]
Can you provide some additional context about what you are trying to achieve?

**Reply by K20shores** [reply]
I only have a cursory understanding. But, one of the projects I work on at NCAR is a giant fortran code generator. For reasons that I won't get into, sometimes we need to parse that fortran. We've mostly gotten by with regex expressions (like [these](https://github.com/NCAR/ccpp-framework/blob/0f416a21050418aaaa5991492802a433a685e2a0/scripts/fortran_tools/parse_fortran_file.py#L32) and [these](https://github.com/NCAR/ccpp-framework/blob/d29a035ddd731c9139d470bb4a8ac018e712a817/scripts/parse_tools/parse_checkers.py#L209)), but this is not really expansive and means that some use cases will not be covered.

It would be fantastic if we could throw a fortran file at fortls in python to scrape the data that we need while [parsing a fortran file](https://github.com/NCAR/ccpp-framework/blob/main/scripts/fortran_tools/parse_fortran_file.py). 

Is that enough information? I can try to explain more if you need it

**Reply by gnikit** [reply]
The answer to what you are asking is somewhat complicated and it depends on what outputs **exactly** you want.
Generally, speaking `fortls` does not offer a high level, user-friendly API to query its internals. There are classes which allow you to parse Fortran syntax quite trivially [API docs link](https://fortls.fortran-lang.org/fortls.parsers.internal.html#fortls.parsers.internal.parser.FortranFile.parse)

```py
from fortls.parsers.internal.parser import FortranFile

file = FortranFile("some_file.f90")
file.load_from_disk()
ast = file.parse()
``` 

> NOTE: APIs might change after https://github.com/fortran-lang/fortls/pull/380

Now whether or not the parsed AST contains the info you want, I have no idea. If you want to share more information with regards to what the outputs of the parsing and what the final result should be, I could probably point you to the right direction.
